### PR TITLE
Fix convolution backward filter cache when using cudnn 7 and above

### PR DESCRIPTION
**Pull request by facug91**

### CODE REVIEW

**Comment by arrufat** [comment]
Oh, did you notice any improvements?

**Comment by facug91** [comment]
I didn't do any performance tests. I was working on another PR idea, when I saw this and decided to fix it. I can do some tests on my machine later, with some of the model training examples, to see how much it was affecting performance. In any case, I think that it will depend a lot on the GPU being used.

**Comment by arrufat** [comment]
I was just curious in case you already had the tests. Don't bother :)

File: dlib/cuda/cudnn_dlibapi.cpp [hunk 1]
```diff
@@ -927,8 +927,8 @@ namespace dlib
             {
                 backward_filters_best_algo = CUDNN_CONVOLUTION_BWD_FILTER_ALGO_0;
             }
-            backward_filters_algo = backward_filters_best_algo;
 #endif
+	        backward_filters_algo = backward_filters_best_algo;
```
Comment (davisking) [inline dlib/cuda/cudnn_dlibapi.cpp:931]:
```suggestion
	    backward_filters_algo = backward_filters_best_algo;
```

**Comment by davisking** [comment]
Oh yeah, thanks for the fix :D

**Comment by arrufat** [comment]
I've just ran the code from here https://github.com/dlibml/dnn and there seems to be a small but consistent improvement. Thank you!

```diff
< alexnet   inference: 2.306 ms (433.700 fps) #params: 62378344 (memory usage: 240.989 MiB) #num convolutions: 5  #num layers: 20
> alexnet   inference: 2.325 ms (430.149 fps) #params: 62378344 (memory usage: 240.989 MiB) #num convolutions: 5  #num layers: 20
< sqznet1.0 inference: 1.158 ms (863.715 fps) #params: 1243176 (memory usage: 34.066 MiB) #num convolutions: 26  #num layers: 65
> sqznet1.0 inference: 1.107 ms (903.597 fps) #params: 1243176 (memory usage: 34.066 MiB) #num convolutions: 26  #num layers: 65
< sqznet1.1 inference: 0.915 ms (1093.389 fps) #params: 1243176 (memory usage: 22.509 MiB) #num convolutions: 26  #num layers: 65
> sqznet1.1 inference: 0.883 ms (1132.119 fps) #params: 1243176 (memory usage: 22.509 MiB) #num convolutions: 26  #num layers: 65
< vggnet11  inference: 4.685 ms (213.460 fps) #params: 132860584 (memory usage: 541.653 MiB) #num convolutions: 8  #num layers: 36
> vggnet11  inference: 4.636 ms (215.704 fps) #params: 132860584 (memory usage: 541.653 MiB) #num convolutions: 8  #num layers: 36
< vggnet13  inference: 5.687 ms (175.834 fps) #params: 133044904 (memory usage: 560.733 MiB) #num convolutions: 10  #num layers: 42
> vggnet13  inference: 5.539 ms (180.529 fps) #params: 133044904 (memory usage: 560.733 MiB) #num convolutions: 10  #num layers: 42
< vggnet16  inference: 7.148 ms (139.898 fps) #params: 138353320 (memory usage: 585.970 MiB) #num convolutions: 13  #num layers: 51
> vggnet16  inference: 6.914 ms (144.638 fps) #params: 138353320 (memory usage: 585.970 MiB) #num convolutions: 13  #num layers: 51
< vggnet19  inference: 8.612 ms (116.116 fps) #params: 143661736 (memory usage: 611.207 MiB) #num convolutions: 16  #num layers: 60
> vggnet19  inference: 8.490 ms (117.786 fps) #params: 143661736 (memory usage: 611.207 MiB) #num convolutions: 16  #num layers: 60
< googlenet inference: 3.255 ms (307.178 fps) #params: 10067368 (memory usage: 60.912 MiB) #num convolutions: 57  #num layers: 196
> googlenet inference: 3.209 ms (311.620 fps) #params: 10067368 (memory usage: 60.912 MiB) #num convolutions: 57  #num layers: 196
< resnet18  inference: 2.149 ms (465.225 fps) #params: 11684008 (memory usage: 58.496 MiB) #num convolutions: 21  #num layers: 70
> resnet18  inference: 2.125 ms (470.594 fps) #params: 11684008 (memory usage: 58.496 MiB) #num convolutions: 21  #num layers: 70
< resnet34  inference: 3.774 ms (264.989 fps) #params: 21784744 (memory usage: 104.237 MiB) #num convolutions: 37  #num layers: 126
> resnet34  inference: 3.692 ms (270.884 fps) #params: 21784744 (memory usage: 104.237 MiB) #num convolutions: 37  #num layers: 126
< resnet50  inference: 5.689 ms (175.787 fps) #params: 25503912 (memory usage: 161.734 MiB) #num convolutions: 53  #num layers: 174
> resnet50  inference: 5.576 ms (179.327 fps) #params: 25503912 (memory usage: 161.734 MiB) #num convolutions: 53  #num layers: 174
< resnet101 inference: 10.070 ms (99.304 fps) #params: 44443816 (memory usage: 266.734 MiB) #num convolutions: 104  #num layers: 344
> resnet101 inference: 9.887 ms (101.142 fps) #params: 44443816 (memory usage: 266.734 MiB) #num convolutions: 104  #num layers: 344
< resnet152 inference: 14.157 ms (70.635 fps) #params: 60041384 (memory usage: 366.618 MiB) #num convolutions: 155  #num layers: 514
> resnet152 inference: 13.865 ms (72.123 fps) #params: 60041384 (memory usage: 366.618 MiB) #num convolutions: 155  #num layers: 514
< darknet19 inference: 2.818 ms (354.807 fps) #params: 20827976 (memory usage: 99.713 MiB) #num convolutions: 18  #num layers: 61
> darknet19 inference: 2.782 ms (359.430 fps) #params: 20827976 (memory usage: 99.713 MiB) #num convolutions: 18  #num layers: 61
< darknet53 inference: 6.982 ms (143.224 fps) #params: 41574216 (memory usage: 211.005 MiB) #num convolutions: 52  #num layers: 181
> darknet53 inference: 6.829 ms (146.437 fps) #params: 41574216 (memory usage: 211.005 MiB) #num convolutions: 52  #num layers: 181
< darknet53csp inference: 6.246 ms (160.090 fps) #params: 27606856 (memory usage: 222.230 MiB) #num convolutions: 72  #num layers: 246
> darknet53csp inference: 6.134 ms (163.018 fps) #params: 27606856 (memory usage: 222.230 MiB) #num convolutions: 72  #num layers: 246
< densenet121 inference: 5.842 ms (171.174 fps) #params: 7895208 (memory usage: 131.729 MiB) #num convolutions: 120  #num layers: 426
> densenet121 inference: 5.745 ms (174.067 fps) #params: 7895208 (memory usage: 131.729 MiB) #num convolutions: 120  #num layers: 426
< densenet169 inference: 8.824 ms (113.326 fps) #params: 13991080 (memory usage: 179.807 MiB) #num convolutions: 168  #num layers: 594
> densenet169 inference: 8.652 ms (115.576 fps) #params: 13991080 (memory usage: 179.807 MiB) #num convolutions: 168  #num layers: 594
< densenet201 inference: 11.874 ms (84.214 fps) #params: 19784872 (memory usage: 244.153 MiB) #num convolutions: 200  #num layers: 706
> densenet201 inference: 11.574 ms (86.403 fps) #params: 19784872 (memory usage: 244.153 MiB) #num convolutions: 200  #num layers: 706
< densenet265 inference: 18.118 ms (55.195 fps) #params: 32939176 (memory usage: 364.239 MiB) #num convolutions: 264  #num layers: 930
> densenet265 inference: 17.739 ms (56.372 fps) #params: 32939176 (memory usage: 364.239 MiB) #num convolutions: 264  #num layers: 930
< densenet161 inference: 13.740 ms (72.781 fps) #params: 28461064 (memory usage: 304.607 MiB) #num convolutions: 160  #num layers: 566
> densenet161 inference: 13.486 ms (74.150 fps) #params: 28461064 (memory usage: 304.607 MiB) #num convolutions: 160  #num layers: 566
< vovnet19s inference: 1.570 ms (636.954 fps) #params: 3159720 (memory usage: 34.040 MiB) #num convolutions: 23  #num layers: 82
> vovnet19s inference: 1.533 ms (652.209 fps) #params: 3159720 (memory usage: 34.040 MiB) #num convolutions: 23  #num layers: 82
< vovnet19  inference: 3.121 ms (320.447 fps) #params: 11193000 (memory usage: 78.430 MiB) #num convolutions: 23  #num layers: 82
> vovnet19  inference: 3.053 ms (327.541 fps) #params: 11193000 (memory usage: 78.430 MiB) #num convolutions: 23  #num layers: 82
< vovnet27s inference: 1.968 ms (508.108 fps) #params: 3643560 (memory usage: 40.197 MiB) #num convolutions: 29  #num layers: 100
> vovnet27s inference: 1.950 ms (512.878 fps) #params: 3643560 (memory usage: 40.197 MiB) #num convolutions: 29  #num layers: 100
< vovnet27  inference: 4.123 ms (242.565 fps) #params: 13136552 (memory usage: 94.466 MiB) #num convolutions: 29  #num layers: 100
> vovnet27  inference: 4.019 ms (248.798 fps) #params: 13136552 (memory usage: 94.466 MiB) #num convolutions: 29  #num layers: 100
< vovnet39  inference: 6.111 ms (163.652 fps) #params: 26184360 (memory usage: 149.374 MiB) #num convolutions: 45  #num layers: 154
> vovnet39  inference: 5.943 ms (168.253 fps) #params: 26184360 (memory usage: 149.374 MiB) #num convolutions: 45  #num layers: 154
< vovnet57  inference: 8.646 ms (115.664 fps) #params: 42441384 (memory usage: 220.106 MiB) #num convolutions: 66  #num layers: 226
> vovnet57  inference: 8.490 ms (117.779 fps) #params: 42441384 (memory usage: 220.106 MiB) #num convolutions: 66  #num layers: 226
< vovnet99  inference: 14.999 ms (66.672 fps) #params: 70412940 (memory usage: 367.421 MiB) #num convolutions: 115  #num layers: 394
> vovnet99  inference: 14.683 ms (68.104 fps) #params: 70412940 (memory usage: 367.421 MiB) #num convolutions: 115  #num layers: 394
< repvgg_a0  inference: 1.474 ms (678.503 fps) #params: 8309384 (memory usage: 38.553 MiB) #num convolutions: 22  #num layers: 46
> repvgg_a0  inference: 1.428 ms (700.354 fps) #params: 8309384 (memory usage: 38.553 MiB) #num convolutions: 22  #num layers: 46
< repvgg_a1  inference: 2.128 ms (469.851 fps) #params: 12789864 (memory usage: 57.846 MiB) #num convolutions: 22  #num layers: 46
> repvgg_a1  inference: 2.089 ms (478.593 fps) #params: 12789864 (memory usage: 57.846 MiB) #num convolutions: 22  #num layers: 46
< repvgg_a2  inference: 3.701 ms (270.175 fps) #params: 25499944 (memory usage: 109.226 MiB) #num convolutions: 22  #num layers: 46
> repvgg_a2  inference: 3.698 ms (270.383 fps) #params: 25499944 (memory usage: 109.226 MiB) #num convolutions: 22  #num layers: 46
< repvgg_b0  inference: 2.506 ms (399.002 fps) #params: 14339048 (memory usage: 66.436 MiB) #num convolutions: 28  #num layers: 58
> repvgg_b0  inference: 2.478 ms (403.548 fps) #params: 14339048 (memory usage: 66.436 MiB) #num convolutions: 28  #num layers: 58
< repvgg_b1  inference: 7.597 ms (131.629 fps) #params: 51829480 (memory usage: 218.019 MiB) #num convolutions: 28  #num layers: 58
> repvgg_b1  inference: 7.479 ms (133.713 fps) #params: 51829480 (memory usage: 218.019 MiB) #num convolutions: 28  #num layers: 58
< repvgg_b2  inference: 11.615 ms (86.098 fps) #params: 80315112 (memory usage: 330.991 MiB) #num convolutions: 28  #num layers: 58
> repvgg_b2  inference: 11.578 ms (86.374 fps) #params: 80315112 (memory usage: 330.991 MiB) #num convolutions: 28  #num layers: 58
< repvgg_b3  inference: 15.836 ms (63.147 fps) #params: 110960872 (memory usage: 452.107 MiB) #num convolutions: 28  #num layers: 58
> repvgg_b3  inference: 15.368 ms (65.069 fps) #params: 110960872 (memory usage: 452.107 MiB) #num convolutions: 28  #num layers: 58


```

**Comment by facug91** [comment]
Just for clarification, this PR should only optimize convolutional network **training**, because the problem was with the chosen backpropagation algorithm.

**Comment by arrufat** [comment]
> Just for clarification, this PR should only optimize convolutional network **training**, because the problem was with the chosen backpropagation algorithm.

Right maybe that just luck or because the GPU had warmed up

**Comment by facug91** [comment]
There could be something else... cuDNN doesn't always use the same algorithms for processing.
With cuDNN 7 and earlier you could be sure that the algorithms were always the same, but since version 8 that has changed, now it tests all the different algorithms available and tells you which worked best in that run. But that can change from one execution to another and can affect overall performance in some cases.

**Pull request merged**